### PR TITLE
 complete lookup method of consul registry and add integration test

### PR DESCRIPTION
**Pull request by Moriadry**
## What is the purpose of the change

 complete lookup method of consul registry and add integration test. see https://github.com/apache/incubator-dubbo/issues/3890

## Brief changelog

XXXXX

##  make lookup method return our immediate provider list, just like what we do in https://github.com/apache/incubator-dubbo/blob/9ce0676908adbb5d073727c47b5a32dda153230e/dubbo-registry/dubbo-registry-zookeeper/src/main/java/org/apache/dubbo/registry/zookeeper/ZookeeperRegistry.java#L195 .
Also check this in ConsulRegistryTest.testLookup()

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by Moriadry** [comment]
@beiwei30 Hi Ian Luo , I found you as the creator of ConsulRegistry, could  you help me review this pr?  I do some fix in it, and add a test to increase the code coverage, and to ensure everything is fine.

File: dubbo-registry/dubbo-registry-consul/pom.xml [hunk 1]
```diff
@@ -36,6 +36,12 @@
             <groupId>com.ecwid.consul</groupId>
             <artifactId>consul-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.pszymczyk.consul</groupId>
+            <artifactId>embedded-consul</artifactId>
+            <version>1.1.0</version>
```
Comment (ralf0131) [inline dubbo-registry/dubbo-registry-consul/pom.xml]: You've specified 2.0.0 in dubbo-dependencies-bom/pom.xml, why it is 1.1.0 here?

File: dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java [hunk 2]
```diff
@@ -154,6 +156,24 @@ public void doUnsubscribe(URL url, NotifyListener listener) {
         notifier.stop();
     }
 
+    @Override
```
Comment (ralf0131) [inline dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: What is the purpose of the change? I could not find any explanation in your pr description.

File: dubbo-registry/dubbo-registry-consul/pom.xml (hunk 1 above)
Comment (Moriadry) [reply dubbo-registry/dubbo-registry-consul/pom.xml]: sorry for the mistake.

File: dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java (hunk 2 above)
Comment (Moriadry) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: If we use lookUp method from parent class: AbstractRegistry, we cannot get our immediate provider list. I choose to implement this method like what we do in https://github.com/apache/incubator-dubbo/blob/9ce0676908adbb5d073727c47b5a32dda153230e/dubbo-registry/dubbo-registry-zookeeper/src/main/java/org/apache/dubbo/registry/zookeeper/ZookeeperRegistry.java#L195 to keep our public method  with consistent meaning. And you can check it in ConsulRegistryTest.testLookup()

**Comment by beiwei30** [comment]
> @beiwei30 Hi Ian, could you help me review this pr? I do some fix in it, I also add a test to increase the code coverage, to ensure everything is fine.

Sorry for the late response. I am on it.

File: dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java [hunk 3]
```diff
@@ -154,6 +156,24 @@ public void doUnsubscribe(URL url, NotifyListener listener) {
...
+    public List<URL> lookup(URL url) {
+        if (url == null) {
+            throw new IllegalArgumentException("lookup url == null");
+        }
+        try {
+            String service = url.getServiceKey();
+            Response<List<HealthService>> result = client.getHealthServices(service, HealthServicesRequest.newBuilder().setTag(SERVICE_TAG).build());
```
Comment (beiwei30) [inline dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: is it possible to reuse **getHealthServices(java.lang.String, long, int)**?

File: dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java [hunk 4]
```diff
@@ -154,6 +156,24 @@ public void doUnsubscribe(URL url, NotifyListener listener) {
...
+            if (result == null || result.getValue() == null || result.getValue().isEmpty()) {
+                return new ArrayList<>();
```
Comment (beiwei30) [inline dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: pls. consider to return Collections.emptyList()

File: dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java (hunk 3 above)
Comment (Moriadry) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: It seems the tcp check never work correctly in consulRegistry. Is there a must that we use tcp check, instead of a http ttl check? I used ttl check, with onlyPass=true, then things went correct.

Comment (Moriadry) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: with this tcp Check  configuration,  client.getHealthService shows the check is critical. What's your opinion sir?

File: dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java (hunk 4 above)
Comment (Moriadry) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: yes, will change this line.

File: dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java (hunk 3 above)
Comment (beiwei30) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: tcp check does work on my side. Using TCP check instead of others does have performance consideration in it.

Comment (beiwei30) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: one more question, why this change is related to TCP check (I guess you mean health check by TCP, am I right?)

Comment (Moriadry) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: yes, consul health check by TCP.  after we register consul service with tcp check, we cannot get  a passing=true HealthService response every time. the ConsulRegistryTest.class now does not involve this logic actually.

Comment (Moriadry) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: I will spend more time on this :)

Comment (Moriadry) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: could u pls check this code(which is implement by me):  https://github.com/baidu/brpc-java/blob/97e6d71d8b83b061384afa9d139a2dc5da841451/brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/ConsulNamingService.java#L126

I believe we need to explicitly healthCheckPass the Check(by checkId) either. Then we can correctly getHealthServices(java.lang.String, long, int) by params: onlyPassing = true

Comment (beiwei30) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: my first version does use **agentCheckPass** instead of **setTcp**.  Not sure why TCP check doesn't work perfectly on your side, would you mind share me your demo app so that I could reproduce. 

If you believe **agentCheckPass** is more robust, would you mind add it back? You could refer my original change: https://github.com/beiwei30/incubator-dubbo/commit/2f828965339176ef76ddb3b91fb74c90d3f3f051 and enhance it. Luckily I didn't delete it from my repo :)

Comment (beiwei30) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: @Moriadry should I merge this pull request first? then you can submit another pull request to follow up with https://github.com/beiwei30/incubator-dubbo/commit/2f828965339176ef76ddb3b91fb74c90d3f3f051. What do you think?

Comment (Moriadry) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: ok, you can merge this, another pr will be provided soon

Comment (beiwei30) [reply dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java]: sure. I will merge it. Thanks for following up. I will add a new samples for consul to test consul TCP check, will ping you once it's ready.
